File: scripts/dev.py
"""Portable entry point; no pre-started bridge. Uses the same JAR-owned runtime on every host."""
import argparse
import json
import os
from pathlib import Path
import shutil
import subprocess
import sys
import zipfile

ROOT=Path(__file__).resolve().parents[1]
sys.path.insert(0,str(ROOT/'bridge'))
from mcandroid_bridge._launch.platforms import Host,capabilities


def compile_core():
    classes=ROOT/'build/portable-classes';classes.mkdir(parents=True,exist_ok=True)
    sources=sorted(p for kind in ('main','test') for p in (ROOT/f'src/{kind}/java/com/zhongbai233/mcandroidphone/core').glob('*.java'))
    for kind in ('main','test'):
        sources.extend(p for p in (ROOT/f'src/{kind}/java/com/zhongbai233/mcandroidphone/phone').glob('*.java')
                       if p.name in ('PhoneGeometry.java','PhonePose.java','PhoneGeometrySelfTest.java'))
    subprocess.run([java_tool('javac'),'-encoding','UTF-8','-d',str(classes),*map(str,sources)],check=True)
    archive=classes/'mcandroidphone/runtime/bridge.zip';archive.parent.mkdir(parents=True,exist_ok=True)
    with zipfile.ZipFile(archive,'w',zipfile.ZIP_DEFLATED) as z:
        for p in (ROOT/'bridge/mcandroid_bridge').rglob('*.py'):z.write(p,p.relative_to(ROOT/'bridge'))
        z.write(ROOT/'LICENSE','licenses/LICENSE')
    return classes


def java_tool(name):
    exe=name+('.exe' if os.name=='nt' else '')
    home=os.environ.get('JAVA_HOME')
    if home and (Path(home)/'bin'/exe).is_file():return str(Path(home)/'bin'/exe)
    found=shutil.which(exe)
    if not found:raise RuntimeError('Install Java 25 and set JAVA_HOME')
    return found


def main():
    parser=argparse.ArgumentParser(description=__doc__)
    parser.add_argument('mode',choices=('check','quick','build','pattern','qemu','bios','runtime-smoke','native-probe'))
    parser.add_argument('--gpu',choices=('virtio','virgl'),default='virtio')
    parser.add_argument('--guest-arch',choices=('amd64','arm64'))
    parser.add_argument('--world-smoke',action='store_true')
    parser.add_argument('--warmup',type=int,default=0)
    parser.add_argument('--set',action='append',default=[],metavar='KEY=VALUE',help='Runtime property; paths stay single argv values')
    args=parser.parse_args();host=Host.current()
    if args.mode=='check':print(json.dumps(capabilities(),indent=2));return
    env=os.environ.copy();env['PYTHONPATH']=str(ROOT/'bridge')
    if args.mode=='native-probe':
        if host.system=='Darwin':
            output=ROOT/'build/native-probe/iosurface-probe';output.parent.mkdir(parents=True,exist_ok=True)
            subprocess.run(['xcrun','clang','-Wno-deprecated-declarations',str(ROOT/'native/macos/iosurface_probe.m'),
                            '-framework','Foundation','-framework','Metal','-framework','IOSurface','-framework','OpenGL','-o',str(output)],check=True)
            result=subprocess.run([str(output)])
            if result.returncode==77: print("GPU capability unavailable on this host; production zero-copy is not validated.")
            elif result.returncode: result.check_returncode()
        elif host.system=='Linux':
            output=ROOT/'build/native-probe/dmabuf-probe';output.parent.mkdir(parents=True,exist_ok=True)
            subprocess.run(['cc',str(ROOT/'native/linux/dmabuf_probe.c'),'-lEGL','-lGL','-o',str(output)],check=True)
            result=subprocess.run([str(output)])
            if result.returncode==77: print("GPU capability unavailable on this host; production zero-copy is not validated.")
            elif result.returncode: result.check_returncode()
        else:raise RuntimeError('Windows GPU validation: scripts/gpu-smoke.py (uses prepared QEMU/Android/ANGLE)')
        return
    if args.mode in ('quick','runtime-smoke'):
        classes=compile_core()
        command=[java_tool('java'),'--enable-native-access=ALL-UNNAMED','-cp',str(classes)]
        if args.mode=='quick':
            for directory in ('bridge/tests','scripts/tests'):
                # Bound each suite and show the blocked thread if a new host exposes a deadlock.
                runner='import faulthandler,runpy; faulthandler.dump_traceback_later(120,exit=True); runpy.run_module("unittest",run_name="__main__"); faulthandler.cancel_dump_traceback_later()'
                subprocess.run([sys.executable,'-u','-c',runner,'discover','-s',directory,'-v'],cwd=ROOT,env=env,check=True)
            subprocess.run([*command,'com.zhongbai233.mcandroidphone.core.CoreSelfTest'],check=True)
            subprocess.run([*command,'com.zhongbai233.mcandroidphone.phone.PhoneGeometrySelfTest'],check=True)
            subprocess.run([*command,'com.zhongbai233.mcandroidphone.core.ManagedRuntimeSelfTest',sys.executable,str(ROOT/'.runtime')],check=True)
        else:
            subprocess.run([*command,'com.zhongbai233.mcandroidphone.core.RuntimeSmoke',sys.executable,str(ROOT/'.runtime'),*args.set],check=True)
        return
    properties={'python':sys.executable,'gpu':args.gpu}
    if args.mode!='build':properties['backend']='pattern' if args.mode=='pattern' else 'qemu'
    if args.mode=='bios':properties.update(bios='true',guestArch='amd64',input='mouse')
    if args.guest_arch:properties['guestArch']=args.guest_arch
    for entry in args.set:
        key,sep,value=entry.partition('=')
        if not sep or key not in ('root','backend','gpu','python','qemu','iso','disk','diskFormat','kernel','initrd',
            'ffmpeg','angle','width','height','density','memory','cpus','accel','guestArch','display','firmware','kernelAppend','input','colorOrder','bios'):
            raise ValueError(f'Unknown runtime override: {key}')
        properties[key]=value
    wrapper=str(ROOT/('gradlew.bat' if os.name=='nt' else 'gradlew'))
    command=([wrapper] if os.name=='nt' else ['sh',wrapper])+['build' if args.mode=='build' else 'runClient']
    command += [f'-PphoneRuntime{k[0].upper()+k[1:]}={v}' for k,v in properties.items()]
    if args.world_smoke:
        command += ['-PphoneWorldSmoke=true',f'-PphoneSmokeWarmupSeconds={args.warmup}']
        if args.mode=='pattern':command.append('-PphonePatternSmoke=true')
        import uuid
        evidence=ROOT/'.runtime/evidence'/str(uuid.uuid4());evidence.mkdir(parents=True)
        command.append('-PphoneSmokeScreenshot='+str(evidence/'phone.png'))
        print('Evidence:',evidence,flush=True)
    subprocess.run(command,cwd=ROOT,check=True)


if __name__=='__main__':main()

File: scripts/Dev.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.zip.*;

/** Java 25 development entry point, also usable without Gradle for core/lifecycle tests. */
class Dev {
    static final Path ROOT=Path.of("").toAbsolutePath(),CLASSES=ROOT.resolve("build/portable-classes");
    static final boolean WINDOWS=System.getProperty("os.name").startsWith("Windows");
    static String java(String name){return Path.of(System.getProperty("java.home"),"bin",name+(WINDOWS?".exe":"")).toString();}
    static void run(List<String> args)throws Exception {int exit=new ProcessBuilder(args).directory(ROOT.toFile()).inheritIO().start().waitFor();if(exit!=0)throw new IOException("Command failed ("+exit+"): "+args.getFirst());}
    static void compile()throws Exception {
        Files.createDirectories(CLASSES);var args=new ArrayList<>(List.of(java("javac"),"-encoding","UTF-8","-d",CLASSES.toString()));
        for(String kind:List.of("main","test")) {
            try(var sources=Files.list(ROOT.resolve("src/"+kind+"/java/com/zhongbai233/mcandroidphone/core"))){sources.filter(p->p.toString().endsWith(".java")).sorted().forEach(p->args.add(p.toString()));}
            try(var sources=Files.list(ROOT.resolve("src/"+kind+"/java/com/zhongbai233/mcandroidphone/phone"))){sources.filter(p->Set.of("PhoneGeometry.java","PhonePose.java","PhoneGeometrySelfTest.java").contains(p.getFileName().toString())).sorted().forEach(p->args.add(p.toString()));}
        }
        try(var sources=Files.list(ROOT.resolve("src/main/java/com/zhongbai233/mcandroidphone/environment"))){sources.filter(p->p.toString().endsWith(".java")).sorted().forEach(p->args.add(p.toString()));}
        run(args);Path jar=CLASSES.resolve("mcandroidphone/runtime/native-guard.jar");Files.createDirectories(jar.getParent());
        try(var zip=new ZipOutputStream(Files.newOutputStream(jar));var files=Files.list(CLASSES.resolve("com/zhongbai233/mcandroidphone/core"))) {
            for(var p:files.filter(p->p.getFileName().toString().matches("(NativeGuard|Json)(\\$.*)?\\.class")).toList()){zip.putNextEntry(new ZipEntry(CLASSES.relativize(p).toString().replace('\\','/')));Files.copy(p,zip);zip.closeEntry();}
        }
    }
    public static void main(String[] args)throws Exception {
        String mode=args.length==0?"pattern":args[0];boolean world=false,android=false,environment=false;String warmup="0";var settings=new LinkedHashMap<String,String>();
        for(int i=1;i<args.length;i++)switch(args[i]) {
            case "--world-smoke"->world=true;
            case "--android-smoke"->{world=true;android=true;}
            case "--environment-smoke"->{world=true;environment=true;settings.put("environment","true");}
            case "--warmup"->warmup=args[++i];
            case "--guest-arch"->settings.put("guestArch",args[++i]);
            case "--gpu","--qemu-gpu"->settings.put("gpu",args[++i]);
            case "--set"->{String value=args[++i];int split=value.indexOf('=');if(split<1)throw new IllegalArgumentException("Expected KEY=VALUE");settings.put(value.substring(0,split),value.substring(split+1));}
            default->throw new IllegalArgumentException("Unknown option: "+args[i]);
        }
        if(android&&!mode.equals("qemu"))throw new IllegalArgumentException("--android-smoke requires qemu and an initialized Android image");
        if(mode.equals("check")){System.out.println("Java "+Runtime.version()+"; "+System.getProperty("os.name")+" "+System.getProperty("os.arch")+"; Python required=false");return;}
        if(mode.equals("quick")||mode.equals("runtime-smoke")) {
            compile();var base=List.of(java("java"),"--enable-native-access=ALL-UNNAMED","-cp",CLASSES.toString());
            if(mode.equals("quick"))for(String test:List.of("core.CoreSelfTest","phone.PhoneGeometrySelfTest","core.EnvironmentSelfTest","core.CameraChannelSelfTest","core.StorageBundleSelfTest","core.SharedSystemDiskSelfTest","core.QmpShutdownSelfTest","core.ManagedRuntimeSelfTest")){var command=new ArrayList<>(base);command.add("com.zhongbai233.mcandroidphone."+test);if(test.endsWith("ManagedRuntimeSelfTest"))command.add(ROOT.resolve(".runtime").toString());run(command);}
            else {var command=new ArrayList<>(base);command.add("com.zhongbai233.mcandroidphone.core.RuntimeSmoke");command.add(ROOT.resolve(".runtime").toString());settings.forEach((k,v)->command.add(k+"="+v));run(command);}return;
        }
        if(mode.equals("native-probe")) {
            if(!System.getProperty("os.name").startsWith("Mac"))throw new IllegalArgumentException("Use platform native probe instructions in docs/portable-validation.md");
            Path out=ROOT.resolve("build/native-probe/iosurface-probe");Files.createDirectories(out.getParent());run(List.of("xcrun","clang","-Wno-deprecated-declarations",ROOT.resolve("native/macos/iosurface_probe.m").toString(),"-framework","Foundation","-framework","Metal","-framework","IOSurface","-framework","OpenGL","-o",out.toString()));int result=new ProcessBuilder(out.toString()).inheritIO().start().waitFor();if(result==77)System.out.println("GPU capability unavailable on this host; no production zero-copy claim.");else if(result!=0)throw new IOException("Native probe failed: "+result);return;
        }
        if(!Set.of("build","pattern","qemu","bios").contains(mode))throw new IllegalArgumentException("Mode: check, quick, build, pattern, qemu, bios, runtime-smoke, native-probe");
        var command=new ArrayList<String>(List.of(java("java"),"-cp",ROOT.resolve("gradle/wrapper/gradle-wrapper.jar").toString(),"org.gradle.wrapper.GradleWrapperMain"));command.add(mode.equals("build")?"build":"runClient");
        if(!mode.equals("build"))settings.putIfAbsent("backend",mode.equals("pattern")?"pattern":"qemu");
        if(mode.equals("bios")){settings.put("bios","true");settings.putIfAbsent("guestArch","amd64");settings.putIfAbsent("input","mouse");}
        settings.forEach((k,v)->command.add("-PphoneRuntime"+Character.toUpperCase(k.charAt(0))+k.substring(1)+"="+v));
        if(world){command.add("-PphoneRuntimeStorage=snapshot");Path evidence=ROOT.resolve(".runtime/evidence/"+UUID.randomUUID());Files.createDirectories(evidence);System.out.println("Evidence: "+evidence);command.addAll(List.of("-PphoneWorldSmoke=true","-PphoneSmokeWarmupSeconds="+warmup,"-PphoneSmokeScreenshot="+evidence.resolve("phone.png")));if(mode.equals("pattern"))command.add("-PphonePatternSmoke=true");}
        if(android)command.add("-PphoneAndroidSmoke=true");
        if(environment)command.add("-PphoneEnvironmentSmoke=true");
        run(command);
    }
}
